File: src/main/java/lesson10/Race.java
package lesson10;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;

public class Race {
    public static final int COMPETITORS_COUNT = 4;
    private CyclicBarrier cyclicBarrier = new CyclicBarrier(COMPETITORS_COUNT);
    private CountDownLatch countDownLatch = new CountDownLatch(COMPETITORS_COUNT);

    public CyclicBarrier getCyclicBarrier() {
        return cyclicBarrier;
    }

    public CountDownLatch getCountDownLatch() {
        return countDownLatch;
    }

    private List<Stage> stages;

    public List<Stage> getStages() {
        return stages;
    }

    public Race(Stage... stages) {
        this.stages = new ArrayList<>(Arrays.asList(stages));
    }

    public void begin() throws InterruptedException {
        System.out.println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Подготовка!!!");
        Car[] cars = new Car[COMPETITORS_COUNT];
        for (int i = 0; i < cars.length; i++) {
            cars[i] = new Car(this, 20 + (int) (Math.random() * 10));
        }
        for (int i = 0; i < cars.length; i++) {
            new Thread(cars[i]).start();
        }
        System.out.println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Гонка началась!!!");
        countDownLatch.await();
        System.out.println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Гонка закончилась!!!");
        Car car = Arrays.stream(cars).min((o1, o2) -> (int) (o1.getFinishTime() - o2.getFinishTime())).get();
        System.out.println(car.getName() + " - победитель");
    }
}

File: src/main/kotlin/lesson10_kt/Race.kt
package lesson10_kt

import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.CyclicBarrier

class Race(vararg stages: Stage?) {
    val stages: List<Stage> = ArrayList(Arrays.asList(*stages))
    val cyclicBarrier = CyclicBarrier(COMPETITORS_COUNT)
    val countDownLatch = CountDownLatch(COMPETITORS_COUNT)

    fun begin() {
        println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Подготовка!!!")
        val cars = arrayOfNulls<Car>(COMPETITORS_COUNT)
        for (i in cars.indices) {
            cars[i] = Car(this, 20 + (Math.random() * 10).toInt())
        }
        for (i in cars.indices) {
            Thread(cars[i]).start()
        }
        println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Гонка началась!!!")
        countDownLatch.await()
        println("ВАЖНОЕ ОБЪЯВЛЕНИЕ >>> Гонка закончилась!!!")
        val winner = cars.minBy { car -> car?.finishTime ?: Long.MAX_VALUE }
        println("Победитель - ${winner?.name}")
    }

    companion object {
        const val COMPETITORS_COUNT: Int = 4
    }
}